### DISCUSSION TITLE
Different Form extractors; same route

### DISCUSSION THREAD

**Discussion by dastardlychimp** [actix/actix-web#1800] [Q&A] [answered]
I'm migrating an old api request into actix web, and I need to keep the same interface. The request allowed two different request structures through `www-form-url-encoded`. I'm struggling to do that in actix.
I've tried deserializing into an enum with the two different variants. The problem there, is that it would deserialize the `Form`'s fields into `String`s instead of expected type.
I also tried two different handlers that have a different `Form` extractor but the same route. Only the first one was used.
And, I tried to use two optional `Form` extractors (Code below). It will correctly parse the first extractor listed, but fail to parse the second.
How can I set up two different `Form` extractors? Is this a bug?

#### Code with two optional `web::Form` extractors
```RUST
use anyhow::anyhow;
use actix_web::{HttpServer, Responder, post, web, App, http::{StatusCode}, error};
use actix_web::middleware::Logger;
use serde::*;
use simple_logger::SimpleLogger;
use log;
use serde_json::json;

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    SimpleLogger::new().init().unwrap();
    
    HttpServer::new(move ||
        App::new()
        .wrap(Logger::default())
        .service(multiple)
    )
        .bind("127.0.0.1:7001")?
        .run()
        .await
}

#[derive(Debug, Deserialize, Serialize)]
struct Test1 {
    test1: i64,
}

#[derive(Debug, Deserialize, Serialize)]
struct Test2 {
    test2: i64,
}

#[post("/multiple")]
async fn multiple(
    form1: Option<web::Form<Test1>>,
    form2: Option<web::Form<Test2>>,
) -> impl Responder {
    log::info!("Got in the multiple function");
    match (form1, form2) {
        (Some(_form1), _) => Ok(format!("Got form1")),
        (_, Some(_form2)) => Ok(format!("Got form2")),
        _ => Err(error::ErrorBadRequest(anyhow!("Netiher form1 or form2"))),
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use actix_web::test;

    async fn test_multiple(form: serde_json::Value) {
        let mut app = test::init_service(App::new().service(multiple)).await;
        
        let req = test::TestRequest::post()
            .uri("/multiple")
            .set_form(&form)
            .to_request();
                
        let mut resp = test::call_service(&mut app, req).await;
        dbg!("{:?}", &resp);

        let bytes = test::load_stream(resp.take_body()).await.unwrap();
        dbg!("{:?}", bytes);

        assert_eq!(resp.status(), StatusCode::OK);
    }

    #[actix_rt::test]
    async fn test_multiple_form1() {
        let form = json!({
            "test1": 1,
        });

        test_multiple(form).await;
    }

    #[actix_rt::test]
    async fn test_multiple_form2() {
        let form = json!({
            "test2": 2,
        });

        test_multiple(form).await;
    }
}
```

#### Cargo test output
```bash
> cargo test
running 2 tests
test test::test_multiple_form1 ... ok
test test::test_multiple_form2 ... FAILED

failures:

---- test::test_multiple_form2 stdout ----
[src/main.rs:60] "{:?}" = "{:?}"
[src/main.rs:60] &resp = 
ServiceResponse HTTP/1.1 400 Bad Request
  headers:
    "content-type": "text/plain; charset=utf-8"
  body: Sized(22)

[src/main.rs:63] "{:?}" = "{:?}"
[src/main.rs:63] bytes = b"Netiher form1 or form2"
thread 'test::test_multiple_form2' panicked at 'assertion failed: `(left == right)`
  left: `400`,
 right: `200`', src/main.rs:65:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Panic in Arbiter thread.


failures:
    test::test_multiple_form2

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

#### Log calling with handler with just `test2` field
> 2020-11-30 07:55:45,763 TRACE [actix_server::worker] Service "actix-web-service-127.0.0.1:7001" is available
2020-11-30 07:55:57,262 TRACE [mio::poll] registering with poller
2020-11-30 07:55:57,262 TRACE [mio::sys::unix::kqueue] registering; token=Token(0); interests=Readable | Writable | Error | Hup
2020-11-30 07:55:57,263 DEBUG [actix_web::extract] Error for Option<T> extractor: Parse error
2020-11-30 07:55:57,263 TRACE [actix_http::h1::decoder] Length read: 8
2020-11-30 07:55:57,263 DEBUG [actix_web::extract] Error for Option<T> extractor: Parse error
2020-11-30 07:55:57,263 INFO  [actix_test] Got in the multiple function
2020-11-30 07:55:57,263 DEBUG [actix_web::middleware::logger] Error in response: Netiher form1 or form2
2020-11-30 07:55:57,263 INFO  [actix_web::middleware::logger] 127.0.0.1:50120 "POST /multiple HTTP/1.1" 400 22 "-" "PostmanRuntime/7.26.8" 0.000531
2020-11-30 07:56:02,266 TRACE [actix_http::h1::dispatcher] Keep-alive timeout, close connection
2020-11-30 07:56:02,266 TRACE [mio::poll] deregistering handle with poller

**Comment by robjtede** [comment, marked as answer]
For this use case, specifically where the body might need to be consumed twice, we adapted the Either type in v3.3 to help: https://docs.rs/actix-web/3.3.1/actix_web/web/enum.Either.html#impl-FromRequest.

Try something like `web::Either<web::Form<Test1>, web::Form<Test2>>`.

**Reply by dastardlychimp** [reply]
Perfect. Thank you!
